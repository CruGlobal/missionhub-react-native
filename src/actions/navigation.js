import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
