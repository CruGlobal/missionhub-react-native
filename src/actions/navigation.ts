/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
