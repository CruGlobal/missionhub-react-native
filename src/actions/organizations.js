import {
  GET_ORGANIZATIONS_CONTACTS_REPORT,
  GET_ORGANIZATION_MEMBERS,
  DEFAULT_PAGE_LIMIT,
  LOAD_ORGANIZATIONS,
  ORGANIZATION_CONTACTS_SEARCH,
  SURVEY_CONTACTS_SEARCH,
  RESET_ORGANIZATION_CONTACTS,
  RESET_SURVEY_CONTACTS,
} from '../constants';
import { timeFilter } from '../utils/filters';
import { organizationSelector } from '../selectors/organizations';

import callApi, { REQUESTS } from './api';

const getOrganizationsQuery = {
  limit: 100,
  include: '',
  filters: {
    descendants: false,
  },
};

export function getMyCommunities() {
  return async dispatch => {
    await dispatch(getMyOrganizations());
    return dispatch(getOrganizationsContactReports());
  };
}

export function getMyOrganizations() {
  return async (dispatch, getState) => {
    const { response: orgs } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATIONS, getOrganizationsQuery),
    );
    const orgOrder = getState().auth.person.user.organization_order;

    if (orgOrder) {
      orgs.sort((a, b) => {
        const aIndex = orgOrder.indexOf(a.id);
        const bIndex = orgOrder.indexOf(b.id);

        if (aIndex === -1) {
          return bIndex === -1 ? 0 : 1;
        }

        if (bIndex === -1) {
          return aIndex === -1 ? 0 : -1;
        }

        return aIndex > bIndex ? 1 : -1;
      });
    }

    dispatch({
      type: LOAD_ORGANIZATIONS,
      orgs,
    });
  };
}

export function getOrganizationsContactReports() {
  return async dispatch => {
    const { response } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATION_INTERACTIONS_REPORT, { period: 'P1W' }),
    );

    dispatch({
      type: GET_ORGANIZATIONS_CONTACTS_REPORT,
      reports: response.map(r => ({
        id: `${r.organization_id}`,
        contactsCount: r.contact_count,
        unassignedCount: r.unassigned_count,
        uncontactedCount: r.uncontacted_count,
      })),
    });
    return response;
  };
}

export function getOrganizationContacts(orgId, name, filters = {}) {
  const query = {
    filters: {
      permissions: 'no_permission',
      organization_ids: orgId,
    },
    include:
      'reverse_contact_assignments,reverse_contact_assignments.organization,organizational_permissions',
  };
  if (name) {
    query.filters.name = name;
  }

  const answerFilters = getAnswersFromFilters(filters);
  if (answerFilters) {
    query.filters.answer_sheets = { answers: answerFilters };
  }
  if (filters.survey) {
    query.filters.answer_sheets = {
      ...(query.filters.answer_sheets || {}),
      survey_ids: filters.survey.id,
    };

    // If there is a survey AND we're filtering by time, apply the time filter to the answer_sheets
    if (filters.time) {
      const dates = timeFilter(filters.time.value);
      query.filters.answer_sheets = {
        ...(query.filters.answer_sheets || {}),
        created_at: [dates.first, dates.last],
      };
    }
  } else {
    // TODO: Enable this when the API supports sorting contacts by `updated_at`
    //   if (filters.time) {
    //     const dates = timeFilter(filters.time.value);
    //     query.filters.updated_at = [dates.first, dates.last];
    //   }
  }
  if (filters.gender) {
    query.filters.genders = filters.gender.id;
  }
  if (filters.archived) {
    query.filters.include_archived = true;
  }
  if (filters.unassigned) {
    query.filters.assigned_tos = 'unassigned';
  }
  if (filters.uncontacted) {
    query.filters.statuses = 'uncontacted';
  }
  if (filters.labels) {
    query.filters.label_ids = filters.labels.id;
  }
  if (filters.groups) {
    query.filters.group_ids = filters.groups.id;
  }

  return async (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });
    const pagination = org.contactPagination || { page: 0, hasNextPage: true };

    const offset = DEFAULT_PAGE_LIMIT * pagination.page;

    query.page = {
      limit: DEFAULT_PAGE_LIMIT,
      offset,
    };

    const { response, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, query),
    );

    return dispatch({
      type: filters.survey
        ? SURVEY_CONTACTS_SEARCH
        : ORGANIZATION_CONTACTS_SEARCH,
      orgId,
      contacts: response,
      query,
      meta,
    });
  };
}

export function reloadOrganizationContacts(orgId, name, filters = {}) {
  return (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });

    if (filters.survey && filters.survey.id) {
      const surveyId = filters.survey.id;
      if (
        org.surveys &&
        org.surveys[surveyId] &&
        org.surveys[surveyId].contactPagination
      ) {
        dispatch(resetContactPagination(orgId, surveyId));
      }
    }

    if (org && org.contactPagination) {
      dispatch(resetContactPagination(orgId));
    }
    return dispatch(getOrganizationContacts(orgId, name, filters));
  };
}

const resetContactPagination = (orgId, surveyId = undefined) => {
  return {
    type: surveyId ? RESET_SURVEY_CONTACTS : RESET_ORGANIZATION_CONTACTS,
    orgId,
    surveyId,
  };
};

//each question/answer filter must be in the URL in the form:
//filters[answers][questionId][]=answerTexts
function getAnswersFromFilters(filters) {
  const arrFilters = Object.keys(filters).map(k => filters[k]);
  const answers = arrFilters.filter(f => f.isAnswer);
  if (answers.length === 0) {
    return null;
  }
  let answerFilters = {};
  answers.forEach(f => {
    answerFilters[f.id] = [f.text];
  });
  return answerFilters;
}

//todo probably should start storing this stuff in Redux
export function getOrganizationMembers(orgId, query = {}) {
  const newQuery = {
    ...query,
    filters: {
      permissions: 'admin,user',
      organization_ids: orgId,
    },
    include: 'organizational_permissions',
  };
  return async dispatch => {
    const { response: members, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, newQuery),
    );

    const memberIds = members.map(m => m.id);
    const reportQuery = {
      people_ids: memberIds.join(','),
      period: 'P1Y',
      organization_ids: orgId,
    };
    const { response: reportResponse } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_INTERACTIONS_REPORT, reportQuery),
    );

    // Get an object with { [key = person_id]: [value = { counts }] }
    const reportsCountObj = reportResponse.reduce(
      (p, n) => ({
        ...p,
        [`${n.person_id}`]: {
          contact_count: n.contact_count,
          uncontacted_count: n.uncontacted_count,
          contacts_with_interaction_count: n.contacts_with_interaction_count,
        },
      }),
      {},
    );
    // Merge the counts into the members array
    const membersWithCounts = members.map(m => ({
      ...m,
      ...(reportsCountObj[m.id] || {}),
    }));

    dispatch({
      type: GET_ORGANIZATION_MEMBERS,
      orgId,
      members: membersWithCounts,
      query: newQuery,
      meta,
    });
    return membersWithCounts;
  };
}

export function getOrganizationMembersNextPage(orgId) {
  return (dispatch, getState) => {
    const { page, hasNextPage } = getState().organizations.membersPagination;
    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = {
      page: {
        limit: DEFAULT_PAGE_LIMIT,
        offset: DEFAULT_PAGE_LIMIT * page,
      },
    };
    return dispatch(getOrganizationMembers(orgId, query));
  };
}

export function addNewContact(data) {
  return (dispatch, getState) => {
    const {
      person: { id: myId },
    } = getState().auth;
    if (!data || !data.firstName) {
      return Promise.reject(
        `Invalid Data from addNewContact: no data or no firstName passed in`,
      );
    }
    let included = [];
    included.push({
      type: 'contact_assignment',
      attributes: {
        assigned_to_id: myId,
        organization_id: data.orgId || undefined,
      },
    });
    if (data.orgId) {
      included.push({
        type: 'organizational_permission',
        attributes: {
          organization_id: data.orgId,
          permission_id: data.orgPermission && data.orgPermission.permission_id,
        },
      });
    }
    if (data.email) {
      included.push({
        type: 'email',
        attributes: { email: data.email },
      });
    }
    if (data.phone) {
      included.push({
        type: 'phone_number',
        attributes: {
          number: data.phone,
          location: 'mobile',
        },
      });
    }
    const bodyData = {
      data: {
        type: 'person',
        attributes: {
          first_name: data.firstName,
          last_name: data.lastName || undefined,
          gender: data.gender || undefined,
        },
      },
      included,
    };
    const query = {};
    return dispatch(callApi(REQUESTS.ADD_NEW_PERSON, query, bodyData));
  };
}
