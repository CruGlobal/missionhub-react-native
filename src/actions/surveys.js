import { DEFAULT_PAGE_LIMIT, GET_ORGANIZATION_SURVEYS } from '../constants';

import callApi, { REQUESTS } from './api';

export function getMySurveys() {
  return dispatch => {
    const query = {
      limit: 100,
      include: '',
    };
    return dispatch(callApi(REQUESTS.GET_SURVEYS, query));
  };
}

export function getOrgSurveys(orgId, query = {}) {
  const newQuery = {
    ...query,
    filters: {
      organization_ids: orgId,
    },
  };
  return async dispatch => {
    const { response, meta } = await dispatch(
      callApi(REQUESTS.GET_GROUP_SURVEYS, newQuery),
    );
    dispatch({
      type: GET_ORGANIZATION_SURVEYS,
      orgId,
      surveys: response,
      query: newQuery,
      meta,
    });
  };
}

export function getOrgSurveysNextPage(orgId) {
  return (dispatch, getState) => {
    const { page, hasNextPage } = getState().organizations.surveysPagination;
    if (!hasNextPage) {
      return Promise.reject('NoMoreData');
    }
    const query = {
      page: {
        limit: DEFAULT_PAGE_LIMIT,
        offset: DEFAULT_PAGE_LIMIT * page,
      },
    };
    return dispatch(getOrgSurveys(orgId, query));
  };
}

export function searchSurveyContacts(text, filters = []) {
  return async dispatch => {
    if (!filters.survey) {
      return Promise.reject('No Survey Specified');
    }

    const query = {
      include: 'person.reverse_contact_assignments',
      filters: createSurveyFilters(filters),
    };

    return await dispatch(callApi(REQUESTS.GET_ANSWER_SHEETS, query));
  };
}

function createSurveyFilters(filters) {
  const answerFilters = getAnswersFromFilters(filters);
  const surveyFilters = {
    survey_ids: filters.survey.id,
    people: {
      organization_ids: filters.organization.id,
    },
  };
  if (answerFilters) {
    surveyFilters.answers = answerFilters;
  }
  if (filters.gender) {
    surveyFilters.people.genders = filters.gender.id;
  }
  if (filters.uncontacted) {
    surveyFilters.people.statuses = 'uncontacted';
  }
  if (filters.unassigned) {
    surveyFilters.people.assigned_tos = 'unassigned';
  }
  surveyFilters.people.include_archived = !!filters.archived;
  if (filters.labels) {
    surveyFilters.people.label_ids = filters.labels.id;
  }
  if (filters.groups) {
    surveyFilters.people.group_ids = filters.groups.id;
  }
  return surveyFilters;
}

//each question/answer filter must be in the URL in the form:
//filters[answers][questionId][]=answerTexts
function getAnswersFromFilters(filters) {
  let answerFilters = {};
  const keys = Object.keys(filters);
  keys.forEach(k => {
    const filter = filters[k];
    if (filter.isAnswer) {
      answerFilters[filter.id] = {
        '': filter.text,
      };
    }
  });
  return answerFilters;
}

export function getSurveyQuestions(surveyId) {
  return async dispatch =>
    await dispatch(callApi(REQUESTS.GET_SURVEY_QUESTIONS, { surveyId }));
}
