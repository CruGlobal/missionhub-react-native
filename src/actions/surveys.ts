import { DEFAULT_PAGE_LIMIT, GET_ORGANIZATION_SURVEYS } from '../constants';
import { REQUESTS } from '../api/routes';

import callApi from './api';

export function getMySurveys() {
  // @ts-ignore
  return dispatch => {
    const query = {
      limit: 100,
      include: '',
    };
    return dispatch(callApi(REQUESTS.GET_SURVEYS, query));
  };
}

// @ts-ignore
export function getOrgSurveys(orgId, query = {}) {
  const newQuery = {
    ...query,
    filters: {
      organization_ids: orgId,
    },
  };
  // @ts-ignore
  return async dispatch => {
    const { response, meta } = await dispatch(
      callApi(REQUESTS.GET_GROUP_SURVEYS, newQuery),
    );
    dispatch({
      type: GET_ORGANIZATION_SURVEYS,
      orgId,
      surveys: response,
      query: newQuery,
      meta,
    });
  };
}

// @ts-ignore
export function getOrgSurveysNextPage(orgId) {
  // @ts-ignore
  return (dispatch, getState) => {
    const { page, hasNextPage } = getState().organizations.surveysPagination;
    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = {
      page: {
        limit: DEFAULT_PAGE_LIMIT,
        offset: DEFAULT_PAGE_LIMIT * page,
      },
    };
    return dispatch(getOrgSurveys(orgId, query));
  };
}

// @ts-ignore
export function getSurveyQuestions(surveyId) {
  // @ts-ignore
  return async dispatch => {
    const { response } = await dispatch(
      callApi(REQUESTS.GET_SURVEY_QUESTIONS, { surveyId }),
    );
    return response;
  };
}

// @ts-ignore
export function getSurveyFilterStats(survey_id) {
  // @ts-ignore
  return async dispatch => {
    const query = {
      survey_id,
    };

    const { response } = await dispatch(
      callApi(REQUESTS.GET_SURVEY_FILTER_STATS, query),
    );
    return response;
  };
}
