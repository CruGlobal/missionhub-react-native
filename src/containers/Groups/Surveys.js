import React, { Component } from 'react';
import { FlatList } from 'react-native';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';

import { Flex, Separator } from '../../components/common';
import GroupSurveyItem from '../../components/GroupSurveyItem';
import LoadMore from '../../components/LoadMore';

import styles from './styles';

@connect()
@translate('groupsSurveys')
class Surveys extends Component {
  handleSelect = survey => {
    LOG('selected survey', survey);
  };

  handleLoadMore = () => {
    LOG('load more');
  };

  render() {
    const { surveys, hasMore } = this.props;
    return (
      <Flex value={1} style={styles.surveys}>
        <FlatList
          data={surveys}
          keyExtractor={i => i.id}
          renderItem={({ item }) => (
            <GroupSurveyItem survey={item} onSelect={this.handleSelect} />
          )}
          ListFooterComponent={
            hasMore ? <LoadMore onPress={this.handleLoadMore} /> : undefined
          }
        />
      </Flex>
    );
  }
}

const mapStateToProps = () => ({
  surveys: [
    {
      id: '1',
      created_at: '2018-05-29T17:02:02Z',
      survey: { title: 'Winter Conference 2018' },
      contactNum: 2,
      uncontactedNum: 3,
      unassignedNum: 5,
    },
    {
      id: '2',
      created_at: '2018-05-28T17:02:02Z',
      survey: { title: 'Freshmen Week 17' },
      contactNum: 50,
      uncontactedNum: 0,
      unassignedNum: 0,
    },
    {
      id: '3',
      created_at: '2018-05-28T17:02:02Z',
      survey: { title: 'Freshmen Week 18' },
      contactNum: 50,
      uncontactedNum: 0,
      unassignedNum: 38,
    },
    {
      id: '4',
      created_at: '2018-05-28T17:02:02Z',
      survey: { title: 'Another Conference for Cru, number 4' },
      contactNum: 50,
      uncontactedNum: 30,
      unassignedNum: 0,
    },
  ],
  hasMore: true,
});

export default connect(mapStateToProps)(Surveys);
