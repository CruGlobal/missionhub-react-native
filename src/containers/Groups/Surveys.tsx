import React, { Component } from 'react';
import { View, FlatList, Linking } from 'react-native';
import { connect } from 'react-redux-legacy';
import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import { RefreshControl } from '../../components/common';
import { refresh, keyExtractorId } from '../../utils/common';
import GroupSurveyItem from '../../components/GroupSurveyItem';
import LoadMore from '../../components/LoadMore';
import { navigatePush } from '../../actions/navigation';
import { getOrgSurveys, getOrgSurveysNextPage } from '../../actions/surveys';
import { refreshCommunity } from '../../actions/organizations';
import { organizationSelector } from '../../selectors/organizations';
import { InfoModal } from '../../components/InfoModal/InfoModal';

import { GROUPS_SURVEY_CONTACTS } from './SurveyContacts';
import styles from './styles';
import OnboardingCard, { GROUP_ONBOARDING_TYPES } from './OnboardingCard';

// @ts-ignore
@withTranslation('groupsSurveys')
class Surveys extends Component {
  state = {
    refreshing: false,
  };

  componentDidMount() {
    // @ts-ignore
    if (this.props.surveys.length === 0) {
      this.load();
    }
  }

  load = () => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    dispatch(refreshCommunity(organization.id));
    return dispatch(getOrgSurveys(organization.id));
  };

  handleRefresh = () => {
    refresh(this, this.load);
  };

  // @ts-ignore
  handleSelect = survey => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    dispatch(
      navigatePush(GROUPS_SURVEY_CONTACTS, {
        organization,
        survey,
      }),
    );
  };

  handleLoadMore = () => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    dispatch(getOrgSurveysNextPage(organization.id));
  };

  // @ts-ignore
  renderItem = ({ item }) => (
    // @ts-ignore
    <GroupSurveyItem survey={item} onSelect={this.handleSelect} />
  );

  // @ts-ignore
  renderHeader = () => <OnboardingCard type={GROUP_ONBOARDING_TYPES.surveys} />;

  render() {
    // @ts-ignore
    const { surveys, pagination, organization, t } = this.props;
    return (
      <View style={styles.pageContainer}>
        <InfoModal
          title={t('movingToWeb')}
          titleStyle={{ maxWidth: 200 }}
          buttonLabel={t('findThemHere').toUpperCase()}
          action={() =>
            Linking.openURL(
              `https://missionhub.com/ministries/${organization.id}/surveyResponses`,
            )
          }
        />
        <FlatList
          data={surveys}
          ListHeaderComponent={this.renderHeader}
          keyExtractor={keyExtractorId}
          style={styles.flatList}
          renderItem={this.renderItem}
          refreshControl={
            <RefreshControl
              refreshing={this.state.refreshing}
              onRefresh={this.handleRefresh}
            />
          }
          ListFooterComponent={
            pagination.hasNextPage ? (
              <LoadMore onPress={this.handleLoadMore} />
            ) : (
              undefined
            )
          }
        />
      </View>
    );
  }
}

// @ts-ignore
Surveys.propTypes = {
  organization: PropTypes.object.isRequired,
};

// @ts-ignore
const mapStateToProps = ({ organizations }, { orgId }) => {
  const organization = organizationSelector({ organizations }, { orgId });

  return {
    // organizations may have _placeholder surveys until the mounting request is completed
    // @ts-ignore
    surveys: (organization.surveys || []).filter(s => !s._placeHolder),
    pagination: organizations.surveysPagination,
    organization,
  };
};

export default connect(mapStateToProps)(Surveys);
