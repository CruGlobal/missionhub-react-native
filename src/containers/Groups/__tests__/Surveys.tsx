import React from 'react';

import Surveys from '../Surveys';
import {
  renderShallow,
  createThunkStore,
  testSnapshotShallow,
} from '../../../../testUtils';
import { navigatePush } from '../../../actions/navigation';
import { getOrgSurveys, getOrgSurveysNextPage } from '../../../actions/surveys';
import { refreshCommunity } from '../../../actions/organizations';
import * as common from '../../../utils/common';

jest.mock('../../../actions/navigation');
jest.mock('../../../actions/surveys');
jest.mock('../../../actions/organizations');

const surveys = [
  {
    id: '1',
    created_at: '2018-08-14T17:02:02Z',
    title: 'Test Survey 1',
    contacts_count: 5,
    unassigned_contacts_count: 5,
    uncontacted_contacts_count: 5,
  },
  {
    id: '2',
    created_at: '2018-08-14T17:02:02Z',
    title: 'Test Survey 2',
    contacts_count: 5,
    unassigned_contacts_count: 5,
    uncontacted_contacts_count: 5,
  },
  {
    id: '3',
    created_at: '2018-08-14T17:02:02Z',
    title: 'Test Survey 3',
    contacts_count: 5,
    unassigned_contacts_count: 5,
    uncontacted_contacts_count: 5,
  },
];
const organization = {
  id: '1',
  name: 'Test Org',
  surveys,
};

// @ts-ignore
let store;

// @ts-ignore
navigatePush.mockReturnValue({ type: 'navigated push' });
// @ts-ignore
getOrgSurveysNextPage.mockReturnValue({ type: 'got org surveys next page' });
// @ts-ignore
getOrgSurveys.mockReturnValue({ type: 'got org surveys' });
// @ts-ignore
refreshCommunity.mockReturnValue({ type: 'refreshed community' });

beforeEach(() => {
  store = createThunkStore({
    organizations: {
      all: [organization],
      surveysPagination: { hasNextPage: true },
    },
  });
});

describe('Surveys', () => {
  const component = <Surveys orgId={organization.id} />;

  it('should render correctly', () => {
    // @ts-ignore
    testSnapshotShallow(component, store);
  });

  it('should mount correctly', () => {
    const store = createThunkStore({
      organizations: {
        all: [
          {
            ...organization,
            surveys: [],
          },
        ],
        surveysPagination: { hasNextPage: true },
      },
    });

    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.componentDidMount();
    expect(refreshCommunity).toHaveBeenCalledWith(organization.id);
    expect(getOrgSurveys).toHaveBeenCalledWith(organization.id);
  });

  it('should handleSelect correctly', () => {
    const buildTrackingResult = { name: 'screen name' };
    // @ts-ignore
    common.buildTrackingObj = jest.fn(() => buildTrackingResult);

    // @ts-ignore
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.handleSelect({ id: '1' });
    expect(navigatePush).toHaveBeenCalled();
  });

  it('should handleLoadMore correctly', () => {
    // @ts-ignore
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.handleLoadMore();
    expect(getOrgSurveysNextPage).toHaveBeenCalled();
  });

  it('should handleRefresh correctly', () => {
    const refresh = jest.fn();
    // @ts-ignore
    common.refresh = refresh;
    // @ts-ignore
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.handleRefresh();
    expect(refresh).toHaveBeenCalled();
  });

  it('renders item', () => {
    // @ts-ignore
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    const renderedItem = instance.renderItem({ item: surveys[0] });
    expect(renderedItem).toMatchSnapshot();
  });

  it('renderHeader match snapshot', () => {
    // @ts-ignore
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    const header = instance.renderHeader();
    expect(header).toMatchSnapshot();
  });
});
