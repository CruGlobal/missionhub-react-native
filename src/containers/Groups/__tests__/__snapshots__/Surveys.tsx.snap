// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surveys renderHeader match snapshot 1`] = `
<OnboardingCard
  type="surveys"
/>
`;

exports[`Surveys renders item 1`] = `
<withI18nextTranslation(GroupSurveyItem)
  onSelect={[Function]}
  survey={
    Object {
      "contacts_count": 5,
      "created_at": "2018-08-14T17:02:02Z",
      "id": "1",
      "title": "Test Survey 1",
      "unassigned_contacts_count": 5,
      "uncontacted_contacts_count": 5,
    }
  }
/>
`;

exports[`Surveys should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <InfoModal
    action={[Function]}
    buttonLabel="FIND THEM HERE"
    title="We are moving Surveys to the web!"
    titleStyle={
      Object {
        "maxWidth": 200,
      }
    }
  />
  <FlatList
    ListFooterComponent={
      <LoadMore
        onPress={[Function]}
      />
    }
    ListHeaderComponent={[Function]}
    data={
      Array [
        Object {
          "contacts_count": 5,
          "created_at": "2018-08-14T17:02:02Z",
          "id": "1",
          "title": "Test Survey 1",
          "unassigned_contacts_count": 5,
          "uncontacted_contacts_count": 5,
        },
        Object {
          "contacts_count": 5,
          "created_at": "2018-08-14T17:02:02Z",
          "id": "2",
          "title": "Test Survey 2",
          "unassigned_contacts_count": 5,
          "uncontacted_contacts_count": 5,
        },
        Object {
          "contacts_count": 5,
          "created_at": "2018-08-14T17:02:02Z",
          "id": "3",
          "title": "Test Survey 3",
          "unassigned_contacts_count": 5,
          "uncontacted_contacts_count": 5,
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    refreshControl={
      <MyRefreshControl
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
